### DISCUSSION TITLE
Possibly usc scittle in a gecko browser or app?

### DISCUSSION THREAD

**Discussion by kangaroolab** [borkdude/scittle#82] [Q&A]
I have a gecko app (based on firefox 60). I tried to import scittle.js but got error:

ReferenceError: globalThis is not defined

my understanding is globalThis is not available with firefox 60. any way to fake it to make it work?

**Comment by borkdude** [comment]
@kangaroolab Firefox 60 is pretty old, current version is 120. But a workaround could be to add:

``` html
<script>var globalThis = window</script>
```

before you load scittle.

**Reply by kangaroolab** [reply]
Thanks.

I tried it with higher version and there is no issue with globalThis. However, when I tried to import scittle.js like below, I got error: TypeError: "toString" is read-only

 Components.utils.import("resource://clojure/scittle.js");

any advice/guidance how to use Scittle with Gecko browser e.g. firefox?

**Reply by borkdude** [reply]
> Components.utils 

This thing is undefined in my locally running Firefox. Any idea how I can test this on a mac? I can try linux maybe later but this is more bothersome for me.

**Comment by kangaroolab** [comment]
I am using one on Linux (linux on my chromebook). here is the doc about components from Mozilla:

https://udn.realityripple.com/docs/Mozilla/Tech/XPCOM/Language_Bindings/Components.utils.import

the DOM model seems to be different between HTML and XHTML. but not sure how tostring is related. 

by looking at scittle.js code, I suspect following is where the error is:
$APP.m.toString=function(){return this.fd};

any insight?

**Reply by borkdude** [reply]
This is probably produced by shadow-cljs which I don't control.
Perhaps you can try to build a simple browser hello world app with shadow-cljs (advanced compiled), try to run it with Gecko and then file an issue with shadow-cljs and then follow up here.

**Comment by kangaroolab** [comment]
Thanks! Last quick question - is there a way to use scittle in a node.js app if it runs 3rd party js module?

**Reply by borkdude** [reply]
You can import scittle like this in a node.js app:

ES6:

``` javascript
import * as scittle from 'scittle/dist/scittle.js';

scittle.default.scittle.core.eval_string(`(prn (+ 1 2 3))`);
```

or CJS:

```
var scittle = require('scittle/dist/scittle.js');
scittle = scittle.scittle.core;
scittle.eval_string(`(prn (+ 1 2 3))`);
```

To use third party libraries you could:

- Define those libraries as globals, and then access them from scittle:

```
var fs = require('fs');
globalThis.fs = fs;
scittle.eval_string(`(js/console.log (js/fs.existsSync "dude"))`);
```

**Reply by borkdude** [reply]
Note that scittle is primarily designed to be used in browsers, so for node.js nbb might be the better fit.
